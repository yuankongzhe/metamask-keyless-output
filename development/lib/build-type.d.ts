import { Infer, Struct } from 'superstruct';

export type Unique<Element extends Struct<any>> = (
  struct: Struct<Infer<Element>[], Infer<Element>>,
  eq?: (a: Infer<Element>, b: Infer<Element>) => boolean,
) => Struct<Infer<Element>[], Infer<Element>>;
