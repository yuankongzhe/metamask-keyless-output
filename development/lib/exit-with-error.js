/**
 * Exit the process with an error message.
 *
 * Note that this should be called before the process ends, but it will not
 * itself end the process. This is because the Node.js documentation strongly
 * advises against calling `process.exit` directly.
 *
 * @param {string} errorMessage - The error message that is causing the non-
 *   zero exit code.
 */
function exitWithError(errorMessage) {
  console.error(errorMessage);
  process.exitCode = 1;
  if (process.env.CI) {
    // For some reason, the CI environment does not exit when process.exitCode
    // is set, or at least doesn't terminate immediately. Something may be
    // keeping the event loop from ending. The result is that sometimes jobs
    // fail because of timeout even though none of the individual tests fail.
    // This is a workaround.
    process.exit();
  }
}

module.exports = { exitWithError };
