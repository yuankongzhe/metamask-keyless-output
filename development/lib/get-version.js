const { version: manifestVersion } = require('../../package.json');
const { loadBuildTypesConfig } = require('./build-type');

/**
 * Get the current version of the MetaMask extension. The base manifest version
 * is modified according to the build type and version.
 *
 * The build version is needed because certain build types (such as beta) may
 * be released multiple times during the release process.
 *
 * @param {string} buildType - The build type.
 * @param {number} buildVersion - The build version.
 * @returns {string} The MetaMask extension version.
 */
function getVersion(buildType, buildVersion) {
  return loadBuildTypesConfig().buildTypes[buildType].isPrerelease === true
    ? `${manifestVersion}-${buildType}.${buildVersion}`
    : manifestVersion;
}

module.exports = { getVersion };
