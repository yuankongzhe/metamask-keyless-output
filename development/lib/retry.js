/**
 * Re-runs the given function until it returns a resolved promise or the number
 * of retries is exceeded, whichever comes first (with an optional delay in
 * between retries).
 *
 * @param {object} args - A set of arguments and options.
 * @param {number} args.retries - The maximum number of times to re-run the
 * function on failure.
 * @param {number} [args.delay] - The amount of time (in milliseconds) to wait in
 * between retries. (Default: 0)
 * @param {string} [args.rejectionMessage] - The message for the rejected promise
 * this function will return in the event of failure. (Default: "Retry limit
 * reached")
 * @param {boolean} [args.retryUntilFailure] - Retries until the function fails.
 * @param {Function} functionToRetry - The function that is run and tested for
 * failure.
 * @returns {Promise<* | null | Error>} a promise that either resolves with one of the following:
 * - If successful, resolves with the return value of functionToRetry.
 * - If functionToRetry fails while retryUntilFailure is true, resolves with null.
 * - Otherwise it is rejected with rejectionMessage.
 */
async function retry(
  {
    retries,
    delay = 0,
    rejectionMessage = 'Retry limit reached',
    retryUntilFailure = false,
  },
  functionToRetry,
) {
  let attempts = 0;
  while (attempts <= retries) {
    if (attempts > 0 && delay > 0) {
      await new Promise((resolve) => setTimeout(resolve, delay));
    }

    try {
      const result = await functionToRetry();
      if (!retryUntilFailure) {
        return result;
      }
    } catch (error) {
      if (error.message === "Exited with code '1'") {
        console.log("retry() received: Exited with code '1'");
      } else {
        console.error('error caught in retry():', error);
      }

      if (attempts < retries) {
        console.log('Ready to retry() again');
      }

      if (retryUntilFailure) {
        return null;
      }
    } finally {
      attempts += 1;
    }
  }

  throw new Error(rejectionMessage);
}

module.exports = { retry };
