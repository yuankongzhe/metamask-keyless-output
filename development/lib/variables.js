/* eslint-disable jsdoc/check-tag-names */
const assert = require('assert');

const DeclaredOnly = Symbol(
  'This variable was declared only without being defined',
);

class Variables {
  /**
   * @type {Map<string, unknown | typeof DeclaredOnly>}
   */
  #definitions = new Map();

  /**
   * @param {Iterable<string>} declarations
   */
  constructor(declarations) {
    for (const declaration of declarations) {
      this.#definitions.set(declaration, DeclaredOnly);
    }
  }

  /**
   * @param {string} key - The name of the variable
   * @throws {TypeError} If there is no definition of a variable.
   */
  get(key) {
    const value = this.getMaybe(key);
    assert(
      value !== DeclaredOnly,
      new TypeError(
        `Tried to access a declared, but not defined environmental variable "${key}"
\tWhy am I seeing this: "${key}" is declared in builds.yml, but had no actual value when we tried loading it.
\tHow do I fix this: You could provide a default value for the variable in builds.yml under "env" property and commit to git. For example:
\t\tenv:
\t\t - ${key}: ''`,
      ),
    );
    return value;
  }

  /**
   * Returns a declared, but maybe not defined variable.
   *
   * @param {string} key - The name of the variable
   * @throws {TypeError} If there was no declaration of the variable.
   * @returns The value, or undefined if the variables wasn't defined.
   */
  getMaybe(key) {
    assert(
      this.isDeclared(key),
      new TypeError(
        `Tried to access an environmental variable "${key}" that wasn't declared in builds.yml
\tWhy am I seeing this: We've made use of new variables be explicit to keep track of all of them in one place
\tHow do I fix this: Adding your variable in builds.yml under "env" property and committing to git will fix this`,
      ),
    );
    return this.#definitions.get(key);
  }

  /**
   * Sets one key
   *
   * @overload
   * @param {string} key
   * @param {unknown} value
   * @returns {void}
   */
  /**
   * @overload
   * @param {Record<string, unknown>} records - Key-Value object
   * @returns {void}
   */
  /**
   * @param {string | Record<string, unknown>} keyOrRecord
   * @param {unknown} value
   * @returns {void}
   */
  set(keyOrRecord, value) {
    if (typeof keyOrRecord === 'object') {
      for (const [key, recordValue] of Object.entries(keyOrRecord)) {
        this.set(key, recordValue);
      }
      return;
    }
    const key = keyOrRecord;
    assert(
      this.isDeclared(key),
      `Tried to modify a variable "${key}" that wasn't declared in builds.yml`,
    );
    assert(value !== DeclaredOnly, `Tried to un-define "${key}" variable`);
    this.#definitions.set(key, value);
  }

  isDeclared(key) {
    return this.#definitions.has(key);
  }

  isDefined(key) {
    return (
      this.#definitions.has(key) && this.#definitions.get(key) !== DeclaredOnly
    );
  }

  [Symbol.iterator] = this.declarations;

  *declarations() {
    yield* this.#definitions.keys();
  }

  *definitions() {
    for (const [key, value] of this.#definitions.entries()) {
      if (value !== DeclaredOnly) {
        yield [key, value];
      }
    }
  }
}

module.exports = { Variables };
