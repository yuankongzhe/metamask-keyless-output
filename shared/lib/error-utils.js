///: BEGIN:ONLY_INCLUDE_IF(desktop)
import browser from 'webextension-polyfill';
///: END:ONLY_INCLUDE_IF
import { memoize } from 'lodash';
import getFirstPreferredLangCode from '../../app/scripts/lib/get-first-preferred-lang-code';
import { fetchLocale, loadRelativeTimeFormatLocaleData } from '../modules/i18n';
///: BEGIN:ONLY_INCLUDE_IF(desktop)
import { renderDesktopError } from '../../ui/pages/desktop-error/render-desktop-error';
import { EXTENSION_ERROR_PAGE_TYPES } from '../constants/desktop';
import { openCustomProtocol } from './deep-linking';
///: END:ONLY_INCLUDE_IF
import switchDirection from './switch-direction';

const _setupLocale = async (currentLocale) => {
  const currentLocaleMessages = currentLocale
    ? await fetchLocale(currentLocale)
    : {};
  const enLocaleMessages = await fetchLocale('en');

  await loadRelativeTimeFormatLocaleData('en');
  if (currentLocale) {
    await loadRelativeTimeFormatLocaleData(currentLocale);
  }

  return { currentLocaleMessages, enLocaleMessages };
};

export const setupLocale = memoize(_setupLocale);

const getLocaleContext = (currentLocaleMessages, enLocaleMessages) => {
  return (key) => {
    let message = currentLocaleMessages[key]?.message;
    if (!message && enLocaleMessages[key]) {
      message = enLocaleMessages[key].message;
    }
    return message;
  };
};

export async function getErrorHtml(
  errorKey,
  supportLink,
  metamaskState,
  ///: BEGIN:ONLY_INCLUDE_IF(desktop)
  err,
  ///: END:ONLY_INCLUDE_IF
) {
  let response, preferredLocale;
  if (metamaskState?.currentLocale) {
    preferredLocale = metamaskState.currentLocale;
    response = await setupLocale(metamaskState.currentLocale);
  } else {
    preferredLocale = await getFirstPreferredLangCode();
    response = await setupLocale(preferredLocale);
  }

  const textDirection = ['ar', 'dv', 'fa', 'he', 'ku'].includes(preferredLocale)
    ? 'rtl'
    : 'auto';

  switchDirection(textDirection);
  const { currentLocaleMessages, enLocaleMessages } = response;
  const t = getLocaleContext(currentLocaleMessages, enLocaleMessages);

  ///: BEGIN:ONLY_INCLUDE_IF(desktop)
  const isDesktopEnabled = metamaskState?.desktopEnabled === true;

  if (isDesktopEnabled) {
    let errorType = EXTENSION_ERROR_PAGE_TYPES.CRITICAL_ERROR;

    if (err?.message.includes('No response from RPC')) {
      errorType = EXTENSION_ERROR_PAGE_TYPES.CONNECTION_LOST;
    }

    return renderDesktopError({
      type: errorType,
      t,
      isHtmlError: true,
    });
  }
  ///: END:ONLY_INCLUDE_IF
  /**
   * The pattern ${errorKey === 'troubleStarting' ? t('troubleStarting') : ''}
   * is neccessary because we we need linter to see the string
   * of the locale keys. If we use the variable directly, the linter will not
   * see the string and will not be able to check if the locale key exists.
   */
  return `
    <div class="critical-error__container">
      <div class="critical-error">
        <div class="critical-error__icon">
          <svg width="24" height="24" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">
            <path d="m443 342l-126-241c-16-32-40-50-65-50-26 0-50 18-66 50l-126 241c-16 30-18 60-5 83 13 23 38 36 71 36l251 0c33 0 58-13 71-36 13-23 11-53-5-83z m-206-145c0-8 6-15 15-15 8 0 14 7 14 15l0 105c0 8-6 15-14 15-9 0-15-7-15-15z m28 182c-1 1-2 2-3 3-1 0-2 1-3 1-1 1-2 1-4 2-1 0-2 0-3 0-2 0-3 0-4 0-2-1-3-1-4-2-1 0-2-1-3-1-1-1-2-2-3-3-4-4-6-9-6-15 0-5 2-11 6-15 1 0 2-1 3-2 1-1 2-2 3-2 1-1 2-1 4-1 2-1 5-1 7 0 2 0 3 0 4 1 1 0 2 1 3 2 1 1 2 2 3 2 4 4 6 10 6 15 0 6-2 11-6 15z"/>
          </svg>
        </div>
        <div>
          <p>
            ${errorKey === 'troubleStarting' ? t('troubleStarting') : ''}
            ${errorKey === 'somethingIsWrong' ? t('somethingIsWrong') : ''}
          </p>
          <div id="critical-error-button" class="critical-error__link critical-error__link-restart">
            ${t('restartMetamask')}
          </div>
          <p class="critical-error__footer">
            <span>${t('stillGettingMessage')}</span>
            <a
              href=${supportLink}
              class="critical-error__link"
              target="_blank"
              rel="noopener noreferrer">
                ${t('sendBugReport')}
              </a>
          </p>
        </div>
      </div>
    </div>
    `;
}

///: BEGIN:ONLY_INCLUDE_IF(desktop)
export const MMD_DOWNLOAD_LINK =
  'https://github.com/MetaMask/metamask-desktop/releases';

function disableDesktop(backgroundConnection) {
  backgroundConnection.disableDesktopError();
}

export function downloadDesktopApp() {
  global.platform.openTab({
    url: MMD_DOWNLOAD_LINK,
  });
}

export function downloadExtension() {
  global.platform.openTab({ url: 'https://metamask.io/' });
}

export function restartExtension() {
  browser.runtime.reload();
}

export function openOrDownloadMMD() {
  openCustomProtocol('metamask-desktop://pair').catch(() => {
    window.open(MMD_DOWNLOAD_LINK, '_blank').focus();
  });
}

export function registerDesktopErrorActions(backgroundConnection) {
  const disableDesktopButton = document.getElementById(
    'desktop-error-button-disable-mmd',
  );
  const restartMMButton = document.getElementById(
    'desktop-error-button-restart-mm',
  );
  const downloadMMDButton = document.getElementById(
    'desktop-error-button-download-mmd',
  );

  const openOrDownloadMMDButton = document.getElementById(
    'desktop-error-button-open-or-download-mmd',
  );

  disableDesktopButton?.addEventListener('click', (_) => {
    disableDesktop(backgroundConnection);
  });

  restartMMButton?.addEventListener('click', (_) => {
    restartExtension();
  });

  downloadMMDButton?.addEventListener('click', (_) => {
    downloadDesktopApp();
  });

  openOrDownloadMMDButton?.addEventListener('click', (_) => {
    openOrDownloadMMD();
  });
}
///: END:ONLY_INCLUDE_IF
