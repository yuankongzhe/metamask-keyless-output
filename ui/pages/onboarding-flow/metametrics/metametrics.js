import React, { useContext } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import Typography from '../../../components/ui/typography/typography';
import {
  TypographyVariant,
  FONT_WEIGHT,
  TEXT_ALIGN,
  TextColor,
  IconColor,
} from '../../../helpers/constants/design-system';
import Button from '../../../components/ui/button';
import { useI18nContext } from '../../../hooks/useI18nContext';
import { setParticipateInMetaMetrics } from '../../../store/actions';
import {
  getFirstTimeFlowTypeRoute,
  getFirstTimeFlowType,
} from '../../../selectors';

import {
  MetaMetricsEventAccountType,
  MetaMetricsEventCategory,
  MetaMetricsEventName,
} from '../../../../shared/constants/metametrics';

import { MetaMetricsContext } from '../../../contexts/metametrics';
import {
  Icon,
  IconName,
  IconSize,
} from '../../../components/component-library';

import Box from '../../../components/ui/box/box';

export default function OnboardingMetametrics() {
  const t = useI18nContext();
  const dispatch = useDispatch();
  const history = useHistory();

  const nextRoute = useSelector(getFirstTimeFlowTypeRoute);
  const firstTimeFlowType = useSelector(getFirstTimeFlowType);

  const trackEvent = useContext(MetaMetricsContext);

  const onConfirm = async () => {
    const [, metaMetricsId] = await dispatch(setParticipateInMetaMetrics(true));
    try {
      trackEvent(
        {
          category: MetaMetricsEventCategory.Onboarding,
          event: MetaMetricsEventName.WalletSetupStarted,
          properties: {
            account_type:
              firstTimeFlowType === 'create'
                ? MetaMetricsEventAccountType.Default
                : MetaMetricsEventAccountType.Imported,
          },
        },
        {
          isOptIn: true,
          metaMetricsId,
          flushImmediately: true,
        },
      );
    } finally {
      history.push(nextRoute);
    }
  };

  const onCancel = async () => {
    await dispatch(setParticipateInMetaMetrics(false));
    history.push(nextRoute);
  };

  return (
    <div
      className="onboarding-metametrics"
      data-testid="onboarding-metametrics"
    >
      <Typography
        variant={TypographyVariant.H2}
        align={TEXT_ALIGN.CENTER}
        fontWeight={FONT_WEIGHT.BOLD}
      >
        {t('onboardingMetametricsTitle')}
      </Typography>
      <Typography
        className="onboarding-metametrics__desc"
        align={TEXT_ALIGN.CENTER}
      >
        {t('onboardingMetametricsDescription')}
      </Typography>
      <Typography
        className="onboarding-metametrics__desc"
        align={TEXT_ALIGN.CENTER}
      >
        {t('onboardingMetametricsDescription2')}
      </Typography>
      <ul>
        <li>
          <Icon
            name={IconName.Check}
            color={IconColor.successDefault}
            marginInlineEnd={3}
          />
          {t('onboardingMetametricsAllowOptOut')}
        </li>
        <li>
          <Icon
            name={IconName.Check}
            color={IconColor.successDefault}
            marginInlineEnd={3}
          />
          {t('onboardingMetametricsSendAnonymize')}
        </li>
        <li>
          <Box>
            <Icon
              marginInlineEnd={2}
              name={IconName.Close}
              size={IconSize.Sm}
              color={IconColor.errorDefault}
            />
            {t('onboardingMetametricsNeverCollect', [
              <Typography
                variant={TypographyVariant.span}
                key="never"
                fontWeight={FONT_WEIGHT.BOLD}
                marginTop={0}
              >
                {t('onboardingMetametricsNeverEmphasis')}
              </Typography>,
            ])}
          </Box>
        </li>
        <li>
          <Box>
            <Icon
              marginInlineEnd={2}
              name={IconName.Close}
              size={IconSize.Sm}
              color={IconColor.errorDefault}
            />
            {t('onboardingMetametricsNeverCollectIP', [
              <Typography
                variant={TypographyVariant.span}
                key="never-collect"
                fontWeight={FONT_WEIGHT.BOLD}
              >
                {t('onboardingMetametricsNeverEmphasis')}
              </Typography>,
            ])}
          </Box>
        </li>
        <li>
          <Box>
            <Icon
              marginInlineEnd={2}
              name={IconName.Close}
              size={IconSize.Sm}
              color={IconColor.errorDefault}
            />
            {t('onboardingMetametricsNeverSellData', [
              <Typography
                variant={TypographyVariant.span}
                key="never-sell"
                fontWeight={FONT_WEIGHT.BOLD}
              >
                {t('onboardingMetametricsNeverEmphasis')}
              </Typography>,
            ])}
          </Box>{' '}
        </li>
      </ul>
      <Typography
        color={TextColor.textAlternative}
        align={TEXT_ALIGN.CENTER}
        variant={TypographyVariant.H6}
        className="onboarding-metametrics__terms"
      >
        {t('onboardingMetametricsDataTerms')}
      </Typography>
      <Typography
        color={TextColor.textAlternative}
        align={TEXT_ALIGN.CENTER}
        variant={TypographyVariant.H6}
        className="onboarding-metametrics__terms"
      >
        {t('onboardingMetametricsInfuraTerms', [
          <a
            href="https://consensys.io/blog/consensys-data-retention-update"
            target="_blank"
            rel="noopener noreferrer"
            key="retention-link"
          >
            {t('onboardingMetametricsInfuraTermsPolicyLink')}
          </a>,
          <a
            href="https://metamask.io/privacy.html"
            target="_blank"
            rel="noopener noreferrer"
            key="privacy-link"
          >
            {t('onboardingMetametricsInfuraTermsPolicy')}
          </a>,
        ])}
      </Typography>

      <div className="onboarding-metametrics__buttons">
        <Button
          data-testid="metametrics-i-agree"
          type="primary"
          large
          onClick={onConfirm}
        >
          {t('onboardingMetametricsAgree')}
        </Button>
        <Button
          data-testid="metametrics-no-thanks"
          type="secondary"
          large
          onClick={onCancel}
        >
          {t('onboardingMetametricsDisagree')}
        </Button>
      </div>
    </div>
  );
}
