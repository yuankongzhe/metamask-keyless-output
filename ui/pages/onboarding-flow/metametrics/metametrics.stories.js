import React from 'react';
import OnboardingMetametrics from './metametrics';

export default {
  title: 'Pages/OnboardingFlow/OnboardingMetametrics',
};

export const DefaultStory = () => <OnboardingMetametrics />;

DefaultStory.storyName = 'Default';
