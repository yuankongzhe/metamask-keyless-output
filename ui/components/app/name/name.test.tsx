import * as React from 'react';
import { NameType } from '@metamask/name-controller';
import configureStore from 'redux-mock-store';
import { renderWithProvider } from '../../../../test/lib/render-helpers';
import { MetaMetricsContext } from '../../../contexts/metametrics';
import {
  MetaMetricsEventCategory,
  MetaMetricsEventName,
} from '../../../../shared/constants/metametrics';
import Name from './name';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => jest.fn(),
}));

const ADDRESS_NO_SAVED_NAME_MOCK = '0xc0ffee254729296a45a3885639ac7e10f9d54977';
const ADDRESS_SAVED_NAME_MOCK = '0xc0ffee254729296a45a3885639ac7e10f9d54979';
const CHAIN_ID_MOCK = '0x1';
const PROPOSED_NAME_MOCK = 'TestProposedName';
const PROPOSED_NAME_2_MOCK = 'TestProposedName2';
const SAVED_NAME_MOCK = 'TestName';
const SOURCE_ID_MOCK = 'TestSourceId';
const SOURCE_ID_2_MOCK = 'TestSourceId2';
const SOURCE_ID_EMPTY_MOCK = 'TestSourceIdEmpty';
const SOURCE_ID_UNDEFINED_MOCK = 'TestSourceIdUndefined';

const STATE_MOCK = {
  metamask: {
    providerConfig: {
      chainId: CHAIN_ID_MOCK,
    },
    names: {
      [NameType.ETHEREUM_ADDRESS]: {
        [ADDRESS_NO_SAVED_NAME_MOCK]: {
          [CHAIN_ID_MOCK]: {
            proposedNames: {
              [SOURCE_ID_MOCK]: [PROPOSED_NAME_MOCK],
              [SOURCE_ID_2_MOCK]: [PROPOSED_NAME_2_MOCK],
              [SOURCE_ID_EMPTY_MOCK]: [],
              [SOURCE_ID_UNDEFINED_MOCK]: undefined,
            },
          },
        },
        [ADDRESS_SAVED_NAME_MOCK]: {
          [CHAIN_ID_MOCK]: {
            proposedNames: { [SOURCE_ID_MOCK]: [PROPOSED_NAME_MOCK] },
            name: SAVED_NAME_MOCK,
          },
        },
      },
    },
  },
};

describe('Name', () => {
  const store = configureStore()(STATE_MOCK);

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders address with no saved name', () => {
    const { container } = renderWithProvider(
      <Name
        type={NameType.ETHEREUM_ADDRESS}
        value={ADDRESS_NO_SAVED_NAME_MOCK}
      />,
      store,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders address with saved name', () => {
    const { container } = renderWithProvider(
      <Name type={NameType.ETHEREUM_ADDRESS} value={ADDRESS_SAVED_NAME_MOCK} />,
      store,
    );

    expect(container).toMatchSnapshot();
  });

  describe('metrics', () => {
    it.each([
      ['saved', ADDRESS_SAVED_NAME_MOCK, true],
      ['not saved', ADDRESS_NO_SAVED_NAME_MOCK, false],
    ])('sends displayed event with %s name', async (_, value, hasPetname) => {
      const trackEventMock = jest.fn();

      renderWithProvider(
        <MetaMetricsContext.Provider value={trackEventMock}>
          <Name type={NameType.ETHEREUM_ADDRESS} value={value} />
        </MetaMetricsContext.Provider>,
        store,
      );

      expect(trackEventMock).toHaveBeenCalledWith({
        event: MetaMetricsEventName.PetnameDisplayed,
        category: MetaMetricsEventCategory.Petnames,
        properties: {
          petname_category: NameType.ETHEREUM_ADDRESS,
          has_petname: hasPetname,
        },
      });
    });
  });
});
