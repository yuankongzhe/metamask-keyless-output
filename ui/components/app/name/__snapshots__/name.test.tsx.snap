// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Name renders address with no saved name 1`] = `
<div>
  <div>
    <div
      class="name name__missing"
    >
      <span
        class="mm-box name__icon mm-icon mm-icon--size-lg mm-box--display-inline-block mm-box--color-inherit"
        style="mask-image: url('./images/icons/question.svg');"
      />
      <p
        class="mm-box mm-text name__value mm-text--body-md mm-box--color-text-default"
      >
        0xC0ffe...54977
      </p>
    </div>
  </div>
</div>
`;

exports[`Name renders address with saved name 1`] = `
<div>
  <div>
    <div
      class="name name__saved"
    >
      <div
        class=""
      >
        <div
          class="identicon"
          style="height: 18px; width: 18px; border-radius: 9px;"
        >
          <div
            style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 18px; height: 18px; display: inline-block; background: rgb(245, 196, 0);"
          >
            <svg
              height="18"
              width="18"
              x="0"
              y="0"
            >
              <rect
                fill="#C8144D"
                height="18"
                transform="translate(-0.9231442102660704 -2.672213349688873) rotate(408.5 9 9)"
                width="18"
                x="0"
                y="0"
              />
              <rect
                fill="#FB1860"
                height="18"
                transform="translate(8.699725589158856 -1.5148649719147114) rotate(402.6 9 9)"
                width="18"
                x="0"
                y="0"
              />
              <rect
                fill="#03505E"
                height="18"
                transform="translate(1.2375522258984861 -14.79073284594546) rotate(308.0 9 9)"
                width="18"
                x="0"
                y="0"
              />
            </svg>
          </div>
        </div>
      </div>
      <p
        class="mm-box mm-text name__name mm-text--body-md mm-box--color-text-default"
      >
        TestName
      </p>
    </div>
  </div>
</div>
`;
